source/util/timer.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "util/timer.h"
+
+#include <sys/resource.h>
+#include <sys/time.h>
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+namespace spvutils {
+
+#if defined(SPIRV_TIMER_ENABLED)
+
+// Print the description of resource types measured by Timer class. If |out| is
+// NULL, it does nothing. Otherwise, it prints resource types. The second is
+// optional and if it is true, the function also prints resource type fields
+// related to memory. Its default is false. In usual, this must be placed before
+// calling Timer::Report() to inform what those fields printed by
+// Timer::Report() indicate.
+void PrintTimerDescription(std::ostream* out, bool measure_mem_usage) {
+  if (out) {
+    *out << std::setw(30) << "PASS name" << std::setw(12) << "CPU time"
+         << std::setw(12) << "WALL time" << std::setw(12) << "USR time"
+         << std::setw(12) << "SYS time";
+    if (measure_mem_usage) {
+      *out << std::setw(12) << "RSS" << std::setw(12) << "Pagefault";
+    }
+    *out << std::endl;
+  }
+}
+
+// Do not change the order of invoking system calls. We want to make CPU/Wall
+// time correct as much as possible. Calling functions to get CPU/Wall time must
+// closely surround the target code of measuring.
+void Timer::Start() {
+  if (report_stream_) {
+    if (getrusage(RUSAGE_SELF, &usage_before_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (clock_gettime(CLOCK_MONOTONIC, &wall_before_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &cpu_before_) == -1) {
+      usage_status_ = kGetrusageFailed;
```

Comment:
Why 'else if'?  If a timer fails, the other timers will not be read.  I think these if() should not be linked with 'else'.

Reply:
Hmm .. the logic here is
1. if some function fails (i.e., the return value is equal to -1), stop and set the error status.
2. else call some other function.

It should be

if (some_function(...) == -1) {
  // it fails ..
  /* report_error */
} else ...

I do not think it is incorrect .. but it would be confusing.

I will change this as
if (some_function(...) == -1) {
  usage_status_ == error;
  return;
}
/* call other functions */

Reply:
Hmm .. it seems that your email is not shown here.

Currently, it prints only when no failure exists.
I will update it so that it prints each type of resource usage if it does not fail.